### ISSUE TITLE
How to address current issues with JavaScript grammar for Go target

### ISSUE THREAD

**Issue by mitar** [antlr/grammars-v4#2177]
See the list of [manual changes one has to do](https://github.com/antlr/grammars-v4/blob/cab9a3798ab30311b5cda257375b6a2e4bb38195/javascript/javascript/Go/README.md) to get to a working JavaScript parser in Go target. I wonder if there is a way to address them (renaming of `this` and not using `*` in lexer base)? Is this something which should be reported upstream to antlr itself?

**Comment by kaby76** [comment]
For actions in the grammar that use "this", the preferred method is to fork a copy of the .g4's in javascript/javascript/ and place them in javascript/javascript/Go, and then modify the grammars there. [Trgen](https://github.com/kaby76/Domemtech.Trash/tree/main/trgen), which is used in CI testing in this repo, copies all the files in [grammars-v4/javascript/javascript](https://github.com/antlr/grammars-v4/tree/master/javascript/javascript), then overlays the files in [grammars-v4/javascript/javascript/Go](https://github.com/antlr/grammars-v4/tree/master/javascript/javascript/Go), then finally overlays instantiated [template files](https://github.com/antlr/grammars-v4/tree/master/_scripts/templates/Go) to create a program to test the grammar.

I am surprised the Antlr4 tool for generating the Go target for the grammar is broken for superClass. If it's broken, the right place would be to make the change in the Antlr4 tool rather than remove the "*" in the generated code.

The rename step for "emptyStatement" to "emptyStatement_" in the [Go-target readme](https://github.com/antlr/grammars-v4/tree/master/javascript/javascript/Go#readme) seems like a problem with the target-specific symbol conflict check in the Antlr4 tool. I just changed 200+ files to remove what I thought would be all the symbol conflicts. If this is a symbol conflict, then the Antlr4 tool needs to be updated and the symbol checked in all grammars in the entire repo.

In lieu of changing the Antlr4 tool, [a modified version of the template build files for the Go target](https://github.com/antlr/grammars-v4/tree/master/_scripts/templates/Go) could be added directly to the target-specific makefile and tester.psm1 files to "sed" the generated files. It isn't pretty, but it would work with a modification to [trgen](https://github.com/kaby76/Domemtech.Trash/tree/main/trgen).

**Comment by studentmain** [comment]
Again, it's time to add language independent semantic predicate syntax in ANTLR...

Since adding new features to ANTLR needs too many works, maybe we can do it ourselves without change ANTLR:
- we have antlr4.g4, then we can find every semantic predicate in a g4 file.
- we have java.g4, then we can even parse those semantic predicates.
- many native code used in it are already simple function calls.
- translate Java function call syntax to other language's seems not too hard.

**Comment by mitar** [comment]
BTW, why does the go target even uses `l` and `p` for receivers? It could simply use `this` as well, `this` is not a keyword in Go. It would look slightly less Go, but it would work out of the box.

The other option might be to have in preamble of the grammar the name for `this`, default being `this`. And then you would have code like `$this.foobar()` where `$this` would be replaced with the value from the preamble.

So it looks like we do not need full parsing here, simply find/replace or placeholders might be enough.

**Comment by kaby76** [comment]
It's worth looking into this, but I don't think we're ready yet. The greater issue is that Antlr itself needs to be replaced. I've been writing a list of requirements for a "new Antlr". I've been thinking about this for a while.

One of the problems with Antlr is that there's no static semantics compiler for it. That's why I been writing Trash for 2 years now. But, the compiler doesn't parse the actions because it's target-language specific, and ANTLRv4Parser/Lexer.g4 parse actions only by matching open and close curly brackets. I have an example program somewhere that parses actions as Java though.

But, we don't know what to support in actions. Is it a general-purpose language? Is it Java? Or, is it DSL that we design? There are string manipulations in lexer actions, tree context checks in parser actions, and looking forward and backward in the token stream in the lexer. There are lexer modes, but this is just a hack for implementing lexing based on context. Should this action language be a totally new language, a highly specific language, or an attribute grammar notation?

**Comment by kaby76** [comment]
Sounds like instead of "p" in the stg code--[here](https://github.com/antlr/antlr4/blob/master/tool/resources/org/antlr/v4/tool/templates/codegen/Go/Go.stg#L245) for example--"this" should be used. That would be a good PR, but like to break some grammars. And who knows if it will ever be integrated into Antlr. I never understood why you can't pass a new collection of templates to the Antlr Tool to generate what you want. That's how I designed trgen--you can do whatever the hell you want. Maybe you can pass the templates, but I just don't know how.

**Comment by mitar** [comment]
>  That would be a good PR, but likely to break some grammars.

We can make it backwards compatible by doing:

```
func (p *<parser.name>) Sempred(localctx antlr.RuleContext, ruleIndex, predIndex int) bool {
  this := p
  ...
```

Make it available under both names. And same for lexer. :-)

> and ANTLRv4Parser/Lexer.g4 parse actions only by matching open and close curly brackets. 

I think this is OK. What is missing in my view is just that you could specify actions per language. So that I could say something like:

```
    | singleExpression [Go]{p.notLineTerminator()}[default]{this.notLineTerminator()}? '++'                     # PostIncrementExpression
```

**Comment by mitar** [comment]
I realized the approach with `this := p` will be backwards incompatible, too. Because Go fails to compile if declared variable is not used. So it will break any way or the other. I think changing `p` to `this` might be the simplest way, but it would be backwards incompatible. What do others think?

**Comment by kaby76** [comment]
Assign to "_". https://stackoverflow.com/questions/21743841/how-to-avoid-annoying-error-declared-and-not-used.

I think the way to go forward is multi pronged. 
* Rename emptyStatement to emptyStatement_ across the board, all grammars in repo.
* Fork a copy of the javascript/javascript/ grammar for Go to rename "this" to "p".
* For CI, update the build templates to get rid of "*" after generating the code.
* Make a PR in antlr/antlr4 for this/p/l in the Go templates and "*" problem.
* Add parsing of actions to Trash compiler. There are tools already to edit and generate the code using Xpath.

**Comment by mitar** [comment]
> Assign to "_"

Nice trick.

> Fork a copy of the javascript/javascript/ grammar for Go to rename "this" to "p".

**Comment by mitar** [comment]
I made: https://github.com/antlr/antlr4/pull/3174

**Comment by studentmain** [comment]
> But, we don't know what to support in actions. Is it a general-purpose language?

I built a prototype. Now I meet this problem.

To keep backward compatibility, and to avoid write a general purpose Java to whatever language compiler. I'd like only support a tiny subset of Java, just enough to call member functions with primitive type's literal.

**Comment by kaby76** [comment]
> > Fork a copy of the javascript/javascript/ grammar for Go to rename "this" to "p".
> 
> But then one has to maintain two versions of the grammar. This is really bad.

Of course. The issue is understood, but a great solution is not possible in the near term. Unfortunately, a release of Antlr is made at the end of the calendar year, with PRs sometimes delayed up to a year before being integrated--if ever!

In the meanwhile, offering nothing is not a great solution either. That's because people will look at the grammar in the repo, then try to run it for a target that we know doesn't work (see [skip lists](https://github.com/antlr/grammars-v4/tree/master/_scripts)). That's exactly how you got here, to this issue. There's no placeholder in the grammars that says "this grammar does not work for such and such target" other than what you see in the skip lists. We don't even test the Antlr4cs "target".

People work from the git repo. They aren't going to run a processor on the Java-specific grammar to compute another target-specific grammar. It seems that's too complicated. However, one could write a tool to compute the grammar for other targets then check that in.

Forking the grammar is a solution--whether it is hand-written, or automatically generated, then checked in--will allow anyone to pick up the grammar and work with it. So, @studentmain's solution is worth investigating.

**Comment by studentmain** [comment]
> However, I'm not sure how far this will go because it will involve the conversion of Java to all targets and for the Java runtime.

The tool will only convert member method calling expression. You'll still need write the base classes one by one.
Personally I won't push it any further, .g4 file will be "language independent", but that's all.
A DSL may be a better choice for language independent sempred, but I'm not sure how to build it correctly and it will take too much times.

**Comment by studentmain** [comment]
![图片](https://user-images.githubusercontent.com/80723249/117039882-cca27d80-acf8-11eb-84ed-b0d2b63a3a03.png)

OK, so it can convert Java function call to Python function call. I'll open a repo for it later.

**Comment by studentmain** [comment]
https://github.com/studentmain/SempredTranslation

**Comment by kaby76** [comment]
@studentmain Yes, this is a start. One problem I wish Antlr would have solved are island grammars. Antlr4 has import, but the namespace for the symbols isn't completely separate between these partitions in the input. Ideally, you could then have a combined parse tree for the whole damn input, and then use an XPath expression to gather all those up for code generation/translation. But, again, not something Antlr will ever do. Also, `@members` would have to be collected and translated.

**Comment by kaby76** [comment]
Related to the Go target:
* "emptyStatement" appears in the java/java8/ and java/java9/ grammars. When "emptyStatement" is used, a compilation error ensues. This symbol must be renamed globally, and classified as a "symbol conflict".
* When it does work, the Go parser for java/java/ is slow. Worse, it fails on grammar java/java/ grammar with input java\java\examples\ManyStringsConcat.java with an out-of-memory condition. The input contains a large number of string literals. I haven't determined whether the runtime problem is due to the lexer or the parser. But, the really poor runtime is due to the handling of memory. The Go target is disabled for the Java grammars, but not Apex because there is only one test. I will create a new issue to report the terrible performance of the Go target for java/java/.

**Comment by mitar** [comment]
There is already issue about performance of Go target: https://github.com/antlr/grammars-v4/issues/1596

**Comment by studentmain** [comment]
> Also, `@members` would have to be collected and translated.

Language translation will be way too complex then. Maybe print a warning about using suggested syntax when we found it.

**Comment by mitar** [comment]
@kaby76 I think this has been reported here: https://github.com/antlr/antlr4/issues/2888

**Comment by kaby76** [comment]
> @kaby76 I think this has been reported here: [antlr/antlr4#2888](https://github.com/antlr/antlr4/issues/2888)

Multiple issues here

* The original issue on perf was opened against an unoptimized grammar. Java/java/ is optimized. Java9 examples fails, too.
* The PR to Set class for Go runtime was never merged. The dev process for Antlr tool and runtime is broken.
* CI Github Actions has no way for PR submitter to kill CI build when it goes rogue. Only an admin can kill it. I removed the JS testing for this reason but it doesn't go far enough. I opened an issue here to add a watchdog to the CI tests here so people just adding a new grammar don't kill the build with Go now.

**Comment by studentmain** [comment]
![图片](https://user-images.githubusercontent.com/80723249/117112427-91479380-ad78-11eb-8505-1c8be1d54805.png)

I think I addressed this issue now. @kaby76 Would you like review the code? I want add a build step to generate corresponding g4 file, so we needn't copy and do tiny modify everywhere.

**Comment by kaby76** [comment]
@studentmain Before you make the PR, I have a few questions, basically concerns about how this is going to fit into this repo. Let's start with the obvious:

Is the CI going to check in the generated source for the Go target of javascript/javascript/ so people can just pick up the grammar?

**Comment by studentmain** [comment]
Generated source code will not put in the repo, as it's "generated code".

**Comment by kaby76** [comment]
People have the expectation that the source in this repo can just be picked up and used. If there's no generated code checked in for the Go target, then they'll just use the default grammar, which won't work because of the "this." syntax. How are folks going to get the generated code so they can just write a parser for JS in Go?

What happens if someone comes along and tries adding a Go-specific version of the grammar to the repo? Let's say it gets stuffed in with a large PR unbeknownst to the admin. What's to prevent that?

(Sorry for being so pedantic.)

**Comment by studentmain** [comment]
> How are folks going to get the generated code so they can just write a parser for JS in Go?

Possible solution:
1. Put generated code in repo. Which will make the repo bloat.
2. Pack them as artifact, so one can download it from CI. GitHub will keep artifact for few day/months.
3. Create release and put them into download. Has no limit, can be automatic too. But you see, we have no release now.
4. Do nothing, add document at readme, so they can download tool and convert it by themselves. Require .NET Core.
5. Do nothing, write a online converter webpage instead. Host a webpage is another bigger problem.

> What's to prevent that?

Test will fail if they did thing wrong in ideal situation. The problem is should we allow a handwritten specific version if we have auto generated version.

**Comment by kaby76** [comment]
Well, let's get rid of the non-starters. I don't like doing (3) or (5)--No one has time for all that.

As for (2). Not sure what image you want to give here. Per grammar, for the whole repo? Whole repo + Generated/ plus target/? You'd then have to start explaining what that's all about. So, I'm also not really into that either.

(4) is doc. The problem with that is that most people aren't going to read documentation when they just want to use the grammar. I read it once, then move on. Things change, your program args change. And, it's not automatable.

Maybe you can just add scripts (Bash and Pwsh) into the directory containing the grammar to execute whatever needs to be done to generate source grammars for the CI or for that person who wants to check out that grammar? If the script file exists, execute it as part of the build. Make the first line in the script be something like "# Execute me first to generate grammars for Go, etc."

Also, when your tool generates any file, it should output a file with the words "Generate by such and such tool, version such and such, date s&s, time s&s". If you have that already, never mind.

**Comment by studentmain** [comment]
> As for (2). Not sure what image you want to give here.

Whole repo with generated g4 file in their position (maybe without test cases to reduce size).

> Maybe you can just add scripts (Bash and Pwsh) into the directory containing the grammar

The repo also became bloat if you put same script in many folders. I'd like put it at / or /_scripts/ .
But we may need some trade off between bloated repo and difficult when using grammar.

> it should output a file with the words

Thanks, I forgot that.

**Comment by kaby76** [comment]
I'm hoping we can fit this into the makefiles so I can just say "make" and it'll look at the dependency and run the tool. So, the _script/templates look like they best be updated.  Also, I'm not sure what the .gitignore file will look like since some cases you want to ignore, and others where the grammars are hand-tweaked to not ignore. But, let's give it a try.

I have a tool to fix watchdogging the build so these neverending builds will go away.It's easy to flub up the so-called skip lists. One misspelling and it won't work, and the test will proceed.

**Comment by mitar** [comment]
Renaming `p` and `l` to `this` is not needed anymore as a workaround has just been merged: https://github.com/antlr/antlr4/pull/3174

**Comment by mitar** [comment]
I am not sure how to address the other two issues.

**Comment by studentmain** [comment]
@kaby76 I added some test cases to the translator, you can add it into CI workflow now.
